batcher/sweep_batcher.go

```diff
@@ -0,0 +1,286 @@
+
+// Batcher is a system that is responsible for accepting sweep requests and
+// placing them in appropriate batches. It will spin up new batches as needed.
+type Batcher struct {
+	// batches is a map of batch IDs to the currently active batches.
+	batches map[lntypes.Hash]*Batch
+
+	// SweepReqs is a channel where sweep requests are received.
+	SweepReqs chan *Sweep
+
+	// errChan is a channel where errors are received.
+	errChan chan error
+
+	// completeChan is a channel where batches signal their completion by
+	// providing their ID.
+	completeChan chan lntypes.Hash
+
+	// wallet is the wallet kit client that is used by batches.
+	wallet lndclient.WalletKitClient
+
+	// chainNotifier is the chain notifier client that is used by batches.
+	chainNotifier lndclient.ChainNotifierClient
+
+	// signerClient is the signer client that is used by batches.
+	signerClient lndclient.SignerClient
+
+	// musig2ServerKit includes all the required functionality to collect
+	// and verify signatures by the swap server in order to cooperatively
+	// sweep funds.
+	musig2ServerSign MuSig2SignSweep
+
+	// store includes all the database interactions that are needed by the
+	// batcher and the batches.
+	store BatcherStore
+}
+
+// NewBatcher creates a new Batcher instance.
+func NewBatcher(wallet lndclient.WalletKitClient,
+	chainNotifier lndclient.ChainNotifierClient,
+	signerClient lndclient.SignerClient, musig2ServerSigner MuSig2SignSweep,
+	store BatcherStore) *Batcher {
+
+	return &Batcher{
+		batches:          make(map[lntypes.Hash]*Batch),
+		SweepReqs:        make(chan *Sweep),
+		errChan:          make(chan error),
+		completeChan:     make(chan lntypes.Hash),
+		wallet:           wallet,
+		chainNotifier:    chainNotifier,
+		signerClient:     signerClient,
+		musig2ServerSign: musig2ServerSigner,
+		store:            store,
+	}
+}
+
+// Run starts the batcher and processes incoming sweep requests.
+func (br *Batcher) Run(ctx context.Context) error {
+	// First we fetch all the batches that are not in a confirmed state from
+	// the database. We will then start these batches and add them to our
+	// map of batches.
+	batches, err := br.FetchUnconfirmedBatches(ctx)
+	if err != nil {
+		return err
+	}
+
+	for _, batch := range batches {
+		_, err := br.spinUpBatchFromDB(ctx, *batch)
+		if err != nil {
+			return err
+		}
+	}
+
+	for {
+		select {
+		case sweep := <-br.SweepReqs:
+			err := br.handleSweep(ctx, sweep)
+			if err != nil {
+				return err
+			}
+
+		case id := <-br.completeChan:
+			delete(br.batches, id)
+
+		case err := <-br.errChan:
+			return err
+
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+}
+
+// handleSweep handles a sweep request by either placing it in an existing
+// batch, or by spinning up a new batch.
+func (br *Batcher) handleSweep(ctx context.Context, sweep *Sweep) error {
+	// Check if the sweep is already in a batch. If that is the case, we
+	// provide the sweep to that batch and return.
```

Comment:
Why to we provide the sweep again if it is already included in that batch?

Reply:
In theory here we could just do a noop and return early. I added this update because I didn't want to exclude the scenario where for some reason some of the `sweep` fields changes (maybe the swap adapted to some changes in its state machine)